sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java

```diff
@@ -316,18 +318,31 @@ public final List<? extends FileBasedSource<T>> split(
     }
   }
 
-  /** Report source Lineage. Depend on the number of files, report full file name or only dir. */
-  private void reportSourceLineage(List<Metadata> expandedFiles) {
+  /**
+   * Report source Lineage. Due to the size limit of Beam metrics, report full file name or only dir
+   * depend on the number of files.
+   *
+   * <p>- Number of files<=100, report full file paths;
+   *
+   * <p>- Number of directory<=100, report up to 100 full directories;
+   *
+   * <p>- Otherwise, report top level only
+   */
+  private static void reportSourceLineage(List<Metadata> expandedFiles) {
     if (expandedFiles.size() <= 100) {
       for (Metadata metadata : expandedFiles) {
         FileSystems.reportSourceLineage(metadata.resourceId());
       }
     } else {
+      HashSet<ResourceId> hashSet = new HashSet<>();
       for (Metadata metadata : expandedFiles) {
-        // TODO(yathu) Currently it simply report one level up if num of files exceeded 100.
-        //  Consider more dedicated strategy (e.g. resolve common ancestor) for accurancy, and work
-        //  with metrics size limit.
-        FileSystems.reportSourceLineage(metadata.resourceId().getCurrentDirectory());
+        ResourceId dir = metadata.resourceId().getCurrentDirectory();
+        hashSet.add(dir);
+        if (hashSet.size() > 100) {
+          FileSystems.reportSourceLineage(dir, LineageLevel.TOP_LEVEL);
+          break;
+        }
+        FileSystems.reportSourceLineage(dir);
```

Comment:
Doesnt this mean that when unique dir > 100 you end up reporting first 100 dire path and then when you see 101 you report the top level leading to mix of dir path reporting and bucket reporting which can be confusing. 
I think a consistent approach will be report dir only if unique dirs < 100 for all known files else report bucket only.

Reply:
Still working on the branch actively. Now fixed. Will ask for PTAL when test passes.